src/server/chunk_server/chunk_server_file_service_impl.cc

```diff
@@ -196,16 +319,164 @@ grpc::Status ChunkServerFileServiceImpl::ApplyMutations(
     grpc::ServerContext* context,
     const protos::grpc::ApplyMutationsRequest* request,
     protos::grpc::ApplyMutationsReply* reply) {
-  // TODO(someone): implement the GFS chunk server logic here
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "needs implementation");
+  LOG(INFO) << "Received ApplyMutationsRequest: " << (*request).DebugString();
+
+  *reply->mutable_request() = *request;
+
+  // Ideally we'll use the serialized order of mutations.
+  // But we are currently just sending one apply mutation request.
+  auto& request_header = request->headers(0);
+
+  // Get data from cache and try to make the local write
+  WriteFileChunkReply write_reply;
+  auto status = WriteFileChunkInternal(request_header, &write_reply);
```

Comment:
Technically we need to recycle the caches by calling Remove on the ChunkDataCacheManager, maybe leave a TODO here to remind ourselves.

Reply:
Yeah forgot to do that. Now removing cache entries after write.